Utils/logger.py

```diff
@@ -0,0 +1,230 @@
+import time
+import sys
+import string
+
+
+# noinspection PyMethodMayBeStatic
+class Logger(object):
+    """
+    The Agent's logging assumptions are:
+    For Log, and LogWithPrefix all messages are logged to the
+    self.file_path and to the self.con_path.  Setting either path
+    parameter to None skips that log.  If Verbose is enabled, messages
+    calling the LogIfVerbose method will be logged to file_path yet
+    not to con_path.  Error and Warn messages are normal log messages
+    with the 'ERROR:' or 'WARNING:' prefix added.
+    """
+
+    def __init__(self, filepath, conpath, verbose=False):
+        """
+        Construct an instance of Logger.
+        """
+        self.file_path = filepath
+        self.con_path = conpath
+        self.verbose = verbose
+
+    def throttle_log(self, counter):
+        """
+        Log everything up to 10, every 10 up to 100, then every 100.
+        """
+        return (counter < 10) or ((counter < 100) and ((counter % 10) == 0)) or ((counter % 100) == 0)
+
+    def write_to_file(self, message):
+        """
+        Write 'message' to logfile.
+        """
+        if self.file_path:
+            try:
+                with open(self.file_path, "a") as F:
+                    message = filter(lambda x: x in string.printable, message)
+
+                    # encoding works different for between interpreter version, we are keeping separate implementation
+                    # to ensure backward compatibility
+                    if sys.version_info[0] == 3:
+                        message = ''.join(list(message)).encode('ascii', 'ignore').decode("ascii", "ignore")
+                    elif sys.version_info[0] == 2:
+                        message = message.encode('ascii', 'ignore')
+
+                    F.write(message + "\n")
+            except IOError as e:
+                pass
+
+    def write_to_console(self, message):
```

Comment:
write_to_file and write_to_console are mostly the same method with two different parameters.